src/main/java/com/github/tomakehurst/wiremock/matching/MultiValuePattern.java

```diff
@@ -16,86 +16,61 @@
 package com.github.tomakehurst.wiremock.matching;
 
 import static java.util.Collections.min;
-import static java.util.Collections.singletonList;
 
-import com.fasterxml.jackson.annotation.JsonCreator;
-import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.github.tomakehurst.wiremock.client.WireMock;
 import com.github.tomakehurst.wiremock.http.MultiValue;
-import com.google.common.base.Function;
-import com.google.common.base.Objects;
-import com.google.common.collect.Lists;
 import java.util.Comparator;
 import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-public class MultiValuePattern implements NamedValueMatcher<MultiValue> {
+@JsonDeserialize(using = MultiValuePatternDeserializer.class)
+public abstract class MultiValuePattern implements NamedValueMatcher<MultiValue> {
 
-  private final StringValuePattern valuePattern;
-
-  public MultiValuePattern(StringValuePattern valuePattern) {
-    this.valuePattern = valuePattern;
-  }
-
-  @JsonCreator
   public static MultiValuePattern of(StringValuePattern valuePattern) {
-    return new MultiValuePattern(valuePattern);
+    return new SingleMatchMultiValuePattern(valuePattern);
   }
 
-  public static MultiValuePattern absent() {
-    return new MultiValuePattern(WireMock.absent());
+  public static MultiValuePattern havingExactly(final StringValuePattern... valuePatterns) {
+    if (valuePatterns.length == 0) {
+      return absent();
+    }
+    return new ExactMatchMultiValuePattern(Stream.of(valuePatterns).collect(Collectors.toList()));
   }
 
-  @Override
-  public MatchResult match(MultiValue multiValue) {
-    List<String> values = multiValue.isPresent() ? multiValue.values() : singletonList(null);
-    return getBestMatch(valuePattern, values);
+  public static MultiValuePattern havingExactly(String... values) {
+    if (values.length == 0) {
+      return absent();
+    }
+    return new ExactMatchMultiValuePattern(
+        Stream.of(values).map(EqualToPattern::new).collect(Collectors.toList()));
   }
 
-  @JsonValue
-  public StringValuePattern getValuePattern() {
-    return valuePattern;
+  public static MultiValuePattern including(final StringValuePattern... valuePatterns) {
+    if (valuePatterns.length == 0) {
+      return absent();
+    }
+    return new IncludesMatchMultiValuePattern(
+        Stream.of(valuePatterns).collect(Collectors.toList()));
   }
 
-  @Override
-  public String getName() {
-    return valuePattern.getName();
+  public static MultiValuePattern including(String... values) {
+    if (values.length == 0) {
+      return absent();
+    }
+    return new IncludesMatchMultiValuePattern(
+        Stream.of(values).map(EqualToPattern::new).collect(Collectors.toList()));
   }
 
-  @Override
-  public String getExpected() {
-    return valuePattern.getExpected();
+  public static MultiValuePattern absent() {
```

Comment:
When moved to `WireMock` I guess this should be renamed to something like `noValues()` to avoid colliding with the single-valued `absent()` method.

Reply:
done